crates/bevy_ecs/src/system/query.rs

```diff
@@ -1436,7 +1467,7 @@ pub enum QueryComponentError {
     /// # Example
     ///
     /// ```
-    /// # use bevy_ecs::{prelude::*, system::QueryComponentError};
+    /// # use bevy_ecs::{prelude::*, system::{QueryComponentError, QueryComponentErrorDetail}};
```

Comment:
(comment applies to the whole `QueryComponentError` rather than this specific line.

### Avoid variants all holding the same data

I see that all variants now are `FooBar(QueryComponentErrorDetail)`. Wouldn't it make more sense to convert `QueryComponentError`  into a `struct`?

This would remove a lot of the boilerplate.

```rust
pub enum QueryComponentErrorKind {
    MissingReadAccess,
    MissingWriteAccess,
    // ...
}
pub struct QueryComponentError {
    pub kind: QueryComponentErrorKind,
    pub entity: Entity,
    pub component: &'static str,
    pub query: &'static str,
}
```

Also `QueryComponentErrorKind` is a fairly long name, it starts to trigger my java-factory detector. After a certain identifier length, it becomes difficult to read code. But I think it's fine as is right now.

### Avoid the `&'static str`

Instead of holding the `str` for the component name and query name, it's possible to make `QueryComponentError` generic over `Q` and `C`. It would look like this:

```rust

pub struct QueryComponentError<Q, C> {
    pub kind: QueryComponentErrorKind,
    pub entity: Entity,
    pub _component: PhantomData<fn(C)>,
    pub _query: PhantomData<fn(Q)>,
}
```

Then, you'd use the `type_name` function in the display impl. This way, it avoids bloating the error type (and therefore the result type) with 32 additional bytes, it's basically free until you display the error.

Reply:
First of all thank you for the review!

I want to say that most of my choice were led by "not changing the existing structure too much" : all the "QuerySomethingError" were enums and two of them were already holding the same kind of data in all variants. QuerySingleError in particular was already holding a `&'static str`. I don't have a strong opinion on them and I did think about the stuff you proposed, but didn't note them all down in the PR, thank you for opening the conversation about them.

About changing enum holding struct to struct with a `Kind` enum field:
- It makes sense, but I'm a bit afraid the kind of error would be "buried" in the doc a bit
- Should I do the same to `QuerySingleError`?

About using `PhantomData` instead of `&'static str`:
- This does make runtime a tiny bit faster if you don't use this data I think, and take less memory I guess, but it would probably make compile time a bit longer and binary size a bit heavier because of monomorphization (which Anyhow already does I think but yeah),
- Why the `fn` in the PhantomData? Is it to avoid explicit lifetime?
- Should I restrict Q and C with a trait constraint? What should I use for Q?
- Should I add helper functions to this type for end user to be able to get the string representation of query and component without having to know about Anyhow nor use the Display trait?

About get_mismatches_detail, I'll reply bellow.

Also since @james7132 self-requested a review, can I have your opinion on all of this, I would appreciate a third opinion before making the changes (I'll have to change all the tests and docs and stuff).